src/test/metrics_test.cpp

```diff
@@ -0,0 +1,414 @@
+    ASSERT_NE(run, nullptr);
+
+    metrics::MetricMetadata metadata = {};
+    EXPECT_DEATH({
+        run->AddMetric<metrics::MetricGauge>(metadata);
+    },
+                 "");
+}
+#endif
+
+TEST(MetricsTest, RunAddMultipleMetric)
+{
+    metrics::Manager manager;
+    metrics::Run*    run = manager.AddRun("run");
+    ASSERT_NE(run, nullptr);
+
+    {
+        metrics::MetricMetadata metadata = {};
+        metadata.name                    = "metric_gauge";
+
+        metrics::MetricGauge* metricGauge = run->AddMetric<metrics::MetricGauge>(metadata);
+        EXPECT_NE(metricGauge, nullptr);
+    }
+
+    {
+        metrics::MetricMetadata metadata = {};
+        metadata.name                    = "metric_counter";
+
+        metrics::MetricCounter* metricCounter = run->AddMetric<metrics::MetricCounter>(metadata);
+        EXPECT_NE(metricCounter, nullptr);
+    }
+}
+
+#if defined(PERFORM_DEATH_TESTS)
+TEST(MetricsTest, RunAddDuplicateMetric)
+{
+    metrics::Manager manager;
+    metrics::Run*    run = manager.AddRun("run");
+    ASSERT_NE(run, nullptr);
+
+    metrics::MetricMetadata metadata = {};
+    metadata.name                    = "metric";
+
+    metrics::MetricGauge* metricGauge = run->AddMetric<metrics::MetricGauge>(metadata);
+    EXPECT_NE(metricGauge, nullptr);
+
+    EXPECT_DEATH({
+        metricGauge = run->AddMetric<metrics::MetricGauge>(metadata);
+    },
+                 "");
+
+    EXPECT_DEATH({
+        metrics::MetricCounter* metricCounter = run->AddMetric<metrics::MetricCounter>(metadata);
+    },
+                 "");
+}
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+// Metrics
+////////////////////////////////////////////////////////////////////////////////
+
+TEST_F(MetricsTestFixture, MetricsCounter)
+{
+    metrics::MetricMetadata metadata;
+    metadata.name                  = "counter";
+    metrics::MetricCounter* metric = run->AddMetric<metrics::MetricCounter>(metadata);
+    EXPECT_EQ(metric->Get(), 0U);
+
+    metric->Increment(1U);
+    EXPECT_EQ(metric->Get(), 1U);
+    metric->Increment(4U);
+    EXPECT_EQ(metric->Get(), 5U);
+}
+
+TEST_F(MetricsTestFixture, MetricsGaugeEntries)
+{
+    metrics::MetricMetadata metadata;
+    metadata.name                = "frame_time";
+    metrics::MetricGauge* metric = run->AddMetric<metrics::MetricGauge>(metadata);
+
+    EXPECT_EQ(metric->GetEntriesCount(), 0U);
+    metric->RecordEntry(0.0000, 11.0);
+    EXPECT_EQ(metric->GetEntriesCount(), 1U);
+    metric->RecordEntry(0.0110, 11.7);
+    EXPECT_EQ(metric->GetEntriesCount(), 2U);
+    metric->RecordEntry(0.0227, 12.2);
+    EXPECT_EQ(metric->GetEntriesCount(), 3U);
+    metric->RecordEntry(0.0349, 10.8);
+    EXPECT_EQ(metric->GetEntriesCount(), 4U);
+    metric->RecordEntry(0.0457, 11.1);
+    EXPECT_EQ(metric->GetEntriesCount(), 5U);
+
+    double seconds;
+    double value;
+    metric->GetEntry(0, &seconds, &value);
+    EXPECT_EQ(seconds, 0.0000);
```

Comment:
EXPECT_NEAR? (input values above cannot be represented as doubles (exactly), so might want to add some margins?

Reply:
There are no computation involved, so I expect the values to be exactly the same.

Reply:
What about the values which are not exactly representable as floats? Nor sure if the conversion to the closest representation is defined.

Reply:
Yeah, I doubt we still have those around. :)